core/src/register_data.rs

```diff
@@ -95,56 +95,85 @@ impl<const SIZE: usize, RB: funty::Integral> RegisterData<RB> for ArrayRegisterD
     }
 }
 
-impl<const SIZE: usize, RB> FromIterator<u8> for ArrayRegisterData<SIZE, RB>
+impl<const SIZE: usize, RB> ArrayRegisterData<SIZE, RB>
 where
     RB: funty::Integral,
     RB::Bytes: for<'a> TryFrom<&'a [u8]>,
 {
-    fn from_iter<T: IntoIterator<Item = u8>>(iter: T) -> Self {
-        // Get the iterator. We assume that it is in the same format as the bytes function outputs
+    /// Try to build a [ArrayRegisterData] from an [IntoIterator<Item = u8>]
+    pub fn try_from_iter<I: IntoIterator<Item = u8>>(
+        iter: I,
+    ) -> Result<Self, RegisterDataFromIterError> {
+        use RegisterDataFromIterError::*;
+        // Get the iterator.
         let mut iter = iter.into_iter();
 
-        assert_eq!(
-            iter.next().unwrap(),
-            REGISTER_DATA_IDENTIFIER,
-            "The given iterator is not for register data"
-        );
+        match iter.next() {
+            Some(REGISTER_DATA_IDENTIFIER) => {}
+            Some(id) => return Err(InvalidIdentifier(id)),
+            None => return Err(NotEnoughItems),
+        }
 
         // First the starting number is encoded
-        let starting_register_number =
-            u16::from_le_bytes([iter.next().unwrap(), iter.next().unwrap()]);
+        let starting_register_number = u16::from_le_bytes([
+            iter.next().ok_or(NotEnoughItems)?,
+            iter.next().ok_or(NotEnoughItems)?,
+        ]);
 
         // Second is how many registers there are
-        let register_count = u16::from_le_bytes([iter.next().unwrap(), iter.next().unwrap()]);
+        let register_count = u16::from_le_bytes([
+            iter.next().ok_or(NotEnoughItems)?,
+            iter.next().ok_or(NotEnoughItems)?,
+        ]);
 
         // Create the buffer we're storing the registers in
         let mut registers = ArrayVec::new();
 
         // We process everything byte-by-byte generically so every register has an unknown length
         // So we need to store the bytes temporarily until we have enough to fully read the bytes as a register
         let register_size = core::mem::size_of::<RB>();
+
+        // Check that all register bytes will fit in `registers`
+        let num_register_bytes = register_count as usize * register_size;
+        if num_register_bytes > SIZE {
+            return Err(LengthTooBig(register_count, register_size));
+        }
```

Comment:
The tests fail. I think it's because SIZE here is the amount of registers and not the byte count of the registers

Reply:
@diondokter tests should be fixed now. Thanks for having a look!